### PR TITLE
create-turbo: improve test cases and make them pass

### PR DESCRIPTION
**Pull request by mehulkar**
- Make existing tests pass
- Add GH action to run tests in CI

  Set up a new Github Action that builds the turbo CLI (since that is
  required for the create-turbo tests) and runs the JS tests for that
  package.

  Notable things:

  - Installs yarn and disables corepack as part of the test suite. This
    is the pristine state we want for the tests that exist today. For
    future tests cases that have a different setup, we may need to
    modify this more.

  - Disables Windows tests. This yarn setup on Windows does not seem to
    work, but this gets us one step forward.

### CODE REVIEW

File: .github/workflows/ci-createturbo.yml
Comment (mehulkar) [inline .github/workflows/ci-createturbo.yml:9]: trying to test this here https://github.com/mehulkar/turborepo/pull/1

Comment (mehulkar) [reply .github/workflows/ci-createturbo.yml]: turns GH actions run without merging the upstream branch, so I guess I don't need the test PR!

File: packages/create-turbo/__tests__/cli.test.ts
Comment (mehulkar) [inline packages/create-turbo/__tests__/cli.test.ts:83]: this is not great, but that's what it is right now in CI (and also locally for me). i could add yarn setup to the action steps to make this better, but might be better as a follow up PR

Comment (nathanhammond) [reply packages/create-turbo/__tests__/cli.test.ts:83]: Does this fix itself with `corepack` being enabled globally?

Comment (mehulkar) [reply packages/create-turbo/__tests__/cli.test.ts:83]: I don't think that will affect this. It won't install yarn for us unless we tell it to. This output just reflects the state of the system when the CLI runs:

https://github.com/vercel/turborepo/blob/655f73ed152943e26aee6cef0fa17f11984ccc01/packages/create-turbo/src/index.ts#L118-L129

Comment (nathanhammond) [reply packages/create-turbo/__tests__/cli.test.ts:83]: Right, but it determines what versions are available by running `yarnpkg -v` which is a shim that will exist of `corepack` is enabled.

File: packages/create-turbo/__tests__/cli.test.ts
Comment (nathanhammond) [inline packages/create-turbo/__tests__/cli.test.ts:95]: We should be able to fix the `(not installed)` before landing this.

`yarn` is actually running, but it is being marked as not installed because `corepack` is erroring since the repository is configured for `pnpm`.

**Review by nathanhammond — CHANGES_REQUESTED** [review]
This will require changes to https://github.com/vercel/turborepo/blob/65f23138861483b3e13889824a91b0f9323ad63b/packages/create-turbo/src/shouldUseYarn.ts (and https://github.com/vercel/turborepo/blob/65f23138861483b3e13889824a91b0f9323ad63b/packages/create-turbo/src/shouldUsePnpm.ts)

File: packages/create-turbo/__tests__/cli.test.ts
Comment (mehulkar) [inline packages/create-turbo/__tests__/cli.test.ts:147]: I would like to make this an `async` function so I can use `await` later, but TS didn't seem to like that, and I'm not sure how to fix.

File: .github/workflows/ci-js.yml
Comment (mehulkar) [inline .github/workflows/ci-js.yml:24]: I'd like to come back to this

File: packages/create-turbo/__tests__/cli.test.ts
Comment (tknickman) [reply packages/create-turbo/__tests__/cli.test.ts:147]: If we moved away from using the `done` callback (which we should do anyway... https://github.com/jest-community/eslint-plugin-jest/blob/main/docs/rules/no-done-callback.md) we could asyncify these.

File: .github/workflows/ci-js.yml
Comment (chris-olszewski) [reply .github/workflows/ci-js.yml:24]: A comment detailing what issues were encountered would be helpful

File: packages/create-turbo/__tests__/cli.test.ts
Comment (mehulkar) [reply packages/create-turbo/__tests__/cli.test.ts:147]: I think I can do that now that i've wrapped the childProcess events into a promise!

File: .github/workflows/ci-js.yml
Comment (mehulkar) [reply .github/workflows/ci-js.yml:24]: It didn't seem to pass in CI and I wanted to isolate the issue and fix. I'm not even sure at this point it's a real problem

File: .github/workflows/ci-js.yml
Comment (nathanhammond) [inline .github/workflows/ci-js.yml:1]: Is this 100% a copy of `ci-go` with the npm section and the `create-turbo` tests added?

File: .github/workflows/ci-js.yml
Comment (nathanhammond) [inline .github/workflows/ci-js.yml:24]: I tried to track down previous runs, got a link? (Don't mind tabling, but I _do_ want to see it.)

File: .github/workflows/ci-js.yml
Comment (nathanhammond) [inline .github/workflows/ci-js.yml:56]: [This would hopefully be deleted at resolution of this issue.](https://github.com/nodejs/corepack/issues/157)

File: packages/create-turbo/__tests__/cli.test.ts
Comment (nathanhammond) [inline packages/create-turbo/__tests__/cli.test.ts:53]: Our most-recent issue with this only showed up during the `install` phase. I'd do a test that actually does `install` for each of our package managers.

File: packages/create-turbo/__tests__/cli.test.ts
Comment (nathanhammond) [inline packages/create-turbo/__tests__/cli.test.ts:62]: Should we localize this for Windows?

File: packages/create-turbo/__tests__/cli.test.ts
Comment (nathanhammond) [inline packages/create-turbo/__tests__/cli.test.ts:79]: Are these easily loopable?

File: .github/workflows/ci-js.yml
Comment (mehulkar) [reply .github/workflows/ci-js.yml:1]: pretty much, but excludes the [`examples`][1] action (terminology?) and excludes [the Go tests][2]

[1]: https://github.com/vercel/turborepo/blob/44ecee64c3c0060f5f7317c5f297b5eb24e09887/.github/workflows/ci-go.yml#L76

[2]: https://github.com/vercel/turborepo/blob/44ecee64c3c0060f5f7317c5f297b5eb24e09887/.github/workflows/ci-go.yml#L70-L74

File: packages/create-turbo/__tests__/cli.test.ts
Comment (mehulkar) [reply packages/create-turbo/__tests__/cli.test.ts:26]: I _think_ so. We could also just _not_ install `corepack` in `ci-js.yml`, but I did it anyway so CI matches what we'll have on dev machines.

Although if/when that issue is resolved, we might _still_ want this so we can represent a real `npx` usecase where a packageManager (or its shim) isn't available and we really do want to show the "not available" case. I don't have a foolproof way forward to address both cases

File: packages/create-turbo/__tests__/cli.test.ts
Comment (mehulkar) [reply packages/create-turbo/__tests__/cli.test.ts:62]: yes absolutely

File: packages/create-turbo/__tests__/cli.test.ts
Comment (mehulkar) [reply packages/create-turbo/__tests__/cli.test.ts:79]: probably! i was trying not to abstract too much here while getting tests passing. Now that I'm adding `install` cases, I may rearrange some more

File: packages/create-turbo/__tests__/cli.test.ts
Comment (nathanhammond) [reply packages/create-turbo/__tests__/cli.test.ts:26]: (That link was mostly to get a pointer to this discussion upstream.)

File: .github/workflows/ci-js.yml
Comment (mehulkar) [reply .github/workflows/ci-js.yml]: https://github.com/vercel/turborepo/runs/7777949925?check_suite_focus=true

I tthink the issue is that `yarn -v` is erroring or the `npm i -g yarn` isn't working. Something for a followup PR

File: packages/create-turbo/__tests__/cli.test.ts
Comment (tknickman) [inline packages/create-turbo/__tests__/cli.test.ts:176]: Can we add a check for:

`>>> Success! Created a new Turborepo at "${DEFAULT_APP_NAME}".`

as well? This would catch that `pnpm` error we had a few weeks ago (the install actually worked, pnpm was just non-zero exiting because it was upset about peers. (which caused us to not show this message and show `install has failed`

**Review by tknickman — COMMENTED** [review]
One comment about an extra test case, but otherwise looks good

File: packages/create-turbo/__tests__/cli.test.ts
Comment (mehulkar) [reply packages/create-turbo/__tests__/cli.test.ts:206]: I can add that, but I wasn't able to reproduce that failure, for what it's worth :/

Comment (mehulkar) [reply packages/create-turbo/__tests__/cli.test.ts:206]: after this PR when tests are running in CI, we can have a bit more confidence about adding new tests if we have regressions again.

**Review by gsoltis — COMMENTED** [review]
I'll defer to @nathanhammond for final LGTM since I think I'm a little out of the loop on the context here. 

In general, this looks fine to me. Just getting these running seems like a step in the right direction. Some things we might want to consider adding, now or in the future

1. displaying the versions of the package managers used, if possible. if an upgrade breaks something, that will help us know
2. are we going to test old and new yarn ("berry")?

**Comment by mehulkar** [comment]
> Some things we might want to consider adding, now or in the future

Good ideas. I think it makes sense to get these running first and add more as needed

> are we going to test old and new yarn ("berry")?

I'd have to think about how to make this possible, since the CI setup (rather than the test case itself) is installing and activating the yarn/npm/etc.

**Comment by gsoltis** [comment]
Test failure? Or is it flaky?

**Review by tknickman — APPROVED** [review]

**Comment by mehulkar** [comment]
> Test failure? Or is it flaky?

I think they are flaky. Looks like it's the Go e2e ones, which shouldn't be affected by this PR. 

```
cli:test: > make test-go
[26](https://github.com/vercel/turborepo/runs/7840795316?check_suite_focus=true#step:10:27)
cli:test: 
[27](https://github.com/vercel/turborepo/runs/7840795316?check_suite_focus=true#step:10:28)
cli:test: go test -race ./...

...[snipped]...

cli:test: Started
[56](https://github.com/vercel/turborepo/runs/7840795316?check_suite_focus=true#step:10:57)
cli:test: --- FAIL: TestStart (0.20s)
[57](https://github.com/vercel/turborepo/runs/7840795316?check_suite_focus=true#step:10:58)
cli:test:     child_test.go:159: process should have exited
```

**Comment by gsoltis** [comment]
@mehulkar I think one of the [js tests](https://github.com/vercel/turborepo/runs/7840795156?check_suite_focus=true#step:11:116) is also failing

**Comment by mehulkar** [comment]
> @mehulkar I think one of the [js tests](https://github.com/vercel/turborepo/runs/7840795156?check_suite_focus=true#step:11:116) is also failing

`npm install`, et al are timing out, even at 50s timeout. Removed them from this PR, and will submit separately to isolate and add logging etc.

**Pull request merged**
